### ISSUE TITLE
"Could not decrypt data with alias: #ServiceName" after Android 13 update

### ISSUE THREAD

**Issue by stetbern** [oblador/react-native-keychain#565]
Hi,

We have had had multiple users report "Could not decrypt data with alias: #ServiceName" after they updated their devices to Android 13. Currently only users of Google Pixel phones are affected. Most of them Pixel 5, but also some Pixel 4/4a and 6.

The error occurs on every app start, which is when we read data from the keychain. I had the chance to look into the Android logs of an affected device:

```
E/android.hardware.keymaster@4.1-service.citadel: FinishOperation : device response error code: INVALID_ARGUMENT
E/keystore2: keystore2::error: In KeystoreOperation::finish
    Caused by:
        0: In finish: KeyMint::finish failed.
        1: Error::Km(ErrorCode(-38))
E/RNKeychainManager: Could not decrypt data with alias: #ServiceName
```

When trying to write data to the keychain, the following error occurs:

```
E/android.hardware.keymaster@4.1-service.citadel: GenerateKey : device response error code: UNSUPPORTED_KEY_SIZE
E/keystore2: keystore2::error: In generate_key.
    Caused by:
        0: While generating Key without explicit attestation key.
        1: Error::Km(ErrorCode(-6))
```
		
I had a look regarding the error code and the only thing I could find was with the generateKey function and the RSA keys here: https://source.android.com/docs/security/keystore/implementer-ref . Maybe somebody knows something more. 

The errors did not come up when the users had Android 12 on their devices. We also tested a fresh install of our app on a newly updated device, but could not reproduce the error. It seems this is connected to the transition from Android 12 to 13.

We store our data with a service and the accesibility ACCESSIBLE.WHEN_UNLOCKED_THIS_DEVICE_ONLY,  all other options as default. On other devices and with other updates (e.g. Android 11->12) this did not cause any issues AFAIK.
We use react-native 0.66.4 and react-native-keychain 8.0.0.

This error is similar to #458 and #466, but as it only appears with Android 13 this might be something different.

**Comment by Meet-JS** [comment]
Thanks a lot. This helped.

**Comment by ilyagru** [comment]
@Meet-JS, could you please share your snippet that works?

Is it just?
```js
await RNKeychain.getGenericPassword({
  storage: STORAGE_TYPE.FB,
});
```

**Comment by Meet-JS** [comment]
@ilyagru My code looks like this:

```
Keychain.setGenericPassword(username, password, {
          storage: STORAGE_TYPE.FB,
        })
```

```
Keychain.getGenericPassword({
          storage: STORAGE_TYPE.FB,
        })
```

**Comment by EJohnF** [comment]
how do you @Meet-JS migrate existing users to FB? 
I see in file `KeychainModule.java` around lines 600-700 that there is some built-in migration. Although I'm not completely sure how it understand what was old storage.

**Comment by yberstad** [comment]
We have seen a sudden increase of this error the last few days, starting to pick up on the 11th of October. Could it be that Samsung is rolling out a security fix that makes this issue worse?
Are any of you experiencing the same?
![276756970-2b281617-258c-4a64-8134-72fc5154fc42](https://github.com/oblador/react-native-keychain/assets/2823722/4852fd3c-2b4c-4a4d-a4c9-5ba0dfc00870)

**Comment by Taraslysak** [comment]
I can confirm that I just got Android system update on my S21 device and this error started occurring

**Comment by AlbertoMeQ** [comment]
Getting the same exact issue:

```
null
                                                                                                    javax.crypto.BadPaddingException
                                                                                                    	at android.security.keystore2.AndroidKeyStoreCipherSpiBase.engineDoFinal(AndroidKeyStoreCipherSpiBase.java:615)
                                                                                                    	at javax.crypto.Cipher.doFinal(Cipher.java:2114)
                                                                                                    	at xb.i.F(CipherStorageKeystoreAesCbc.java:17)
                                                                                                    	at xb.i.E(CipherStorageKeystoreAesCbc.java:3)
                                                                                                    	at xb.i.D(CipherStorageKeystoreAesCbc.java:24)
                                                                                                    	at xb.i.i(CipherStorageKeystoreAesCbc.java:2)
                                                                                                    	at com.oblador.keychain.KeychainModule.decryptToResult(KeychainModule.java:20)
                                                                                                    	at com.oblador.keychain.KeychainModule.decryptCredentials(KeychainModule.java:13)
                                                                                                    	at com.oblador.keychain.KeychainModule.getGenericPassword(KeychainModule.java:83)
                                                                                                    	at com.oblador.keychain.KeychainModule.getGenericPasswordForOptions(KeychainModule.java:5)
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method)
                                                                                                    	at com.facebook.react.bridge.JavaMethodWrapper.invoke(JavaMethodWrapper.java:149)
                                                                                                    	at com.facebook.react.bridge.JavaModuleWrapper.invoke(JavaModuleWrapper.java:148)
                                                                                                    	at com.facebook.jni.NativeRunnable.run(Native Method)
                                                                                                    	at android.os.Handler.handleCallback(Handler.java:942)
                                                                                                    	at android.os.Handler.dispatchMessage(Handler.java:99)
                                                                                                    	at com.facebook.react.bridge.queue.MessageQueueThreadHandler.dispatchMessage(MessageQueueThreadHandler.java:1)
                                                                                                    	at android.os.Looper.loopOnce(Looper.java:226)
                                                                                                    	at android.os.Looper.loop(Looper.java:313)
                                                                                                    	at com.facebook.react.bridge.queue.MessageQueueThreadImpl$4.run(MessageQueueThreadImpl.java:38)
                                                                                                    	at java.lang.Thread.run(Thread.java:1012)
                                                                                                    Caused by: android.security.KeyStoreException: Invalid argument (internal Keystore code: -38 message: In KeystoreOperation::finish
                                                                                                    
                                                                                                    Caused by:
                                                                                                        0: In finish: KeyMint::finish failed.
                                                                                                        1: Error::Km(ErrorCode(-38))) (public error code: 10 internal Keystore code: -38)
                                                                                                    	at android.security.KeyStore2.getKeyStoreException(KeyStore2.java:418)
                                                                                                    	at android.security.KeyStoreOperation.handleExceptions(KeyStoreOperation.java:78)
                                                                                                    	at android.security.KeyStoreOperation.finish(KeyStoreOperation.java:128)
                                                                                                    	at android.security.keystore2.KeyStoreCryptoOperationChunkedStreamer$MainDataStream.finish(KeyStoreCryptoOperationChunkedStreamer.java:228)
                                                                                                    	at android.security.keystore2.KeyStoreCryptoOperationChunkedStreamer.doFinal(KeyStoreCryptoOperationChunkedStreamer.java:181)
                                                                                                    	at android.security.keystore2.AndroidKeyStoreCipherSpiBase.engineDoFinal(AndroidKeyStoreCipherSpiBase.java:609)
                                                                                                    	... 20 more
2024-01-17 13:49:37.902 30164-30222 RNKeychainManager       com.example.ex        E  Could not decrypt data with alias: 
```

Cannot seem to be able to find a fix :/

**Comment by jsite-pl** [comment]
I managed problem when getting same error with same stack trace on Pixel 4 with Android 13, and it get saved value after ~1500ms after use setGenericPassword.

    "react-native": "0.72.4",
    "react-native-keychain": "^8.0.0", // After fix, works on 8.2.0

My initial code:

    export const setGenericSecureValue = async (key: string, value: string, service: string) =>
      await Keychain.setGenericPassword(key, value, { service })
    
    export const getGenericSecureValue = async (key: string, service: string) => {
      try {
        const result = await Keychain.getGenericPassword({
          service,
          ...(Platform.OS !== 'ios' ? { storage: Keychain.STORAGE_TYPE.FB } : {}),
        })
    
        if (!result) {
          return false
        }
    
        return result.password
      } catch (error) {
        return false
      }
    }
    
    
    await setGenericSecureValue('xx','yy','cc');
    await getGenericSecureValue('xx','cc'); // <-- this return error
    setTimeout(async () => {
	    const saved = await getGenericSecureValue('xx','cc'); // <-- this returned correct value
    // do something
    }, 1500)

I fixed that with add storage option to setGenericPassword and getGenericPassword:

    await Keychain.setGenericPassword(key, value, {
	    service,
	    storage: Keychain.STORAGE_TYPE.FB, // <-- Added this
    })
    const result = await await Keychain.getGenericPassword({
        service,
        storage: Keychain.STORAGE_TYPE.FB, // <-- Added this
    })
    
now there is no need setTimeout or other waiting func to get correct value
